webnn/resources/utils.js

```diff
@@ -208,86 +230,159 @@ const assert_array_approx_equals_ulp = (actual, expected, nulp, dataType, descri
+  const description = `test ${name} ${operandType}`;
+  if (typeof expected === 'number') {
+    // for checking a scalar output by matmul 1D x 1D
+    expected = [expected];
+    actual = [actual];
+  }
+  if (metricType === 'ULP') {
+    assert_array_approx_equals_ulp(actual, expected, tolerance, operandType, description);
+  } else if (metricType === 'ATOL') {
+    assert_array_approx_equals(actual, expected, tolerance, description);
+  }
+};
+
+/**
+ * Check computed results with expected data.
+ * @param {String} name - An operation name
  * @param {Object.<String, MLOperand>} namedOutputOperands
  * @param {Object.<MLNamedArrayBufferViews>} outputs - The resources of required outputs
  * @param {Object} resources - Resources used for building a graph
  */
-const checkResults = (opName, operandType, namedOutputOperands, outputs, resources) => {
-  const metricType = PrecisionMetrics[opName] ? Object.keys(PrecisionMetrics[opName])[0] : 'ULP';
-  const description = `test ${opName} ${operandType}`;
-  const tolerance = getPrecisonTolerance(opName, metricType, operandType, resources);
-  if (metricType === 'ULP') {
+const checkResults = (name, namedOutputOperands, outputs, resources) => {
+  const operandType = resources.type;
+  const metricType = PrecisionMetrics[name] ? Object.keys(PrecisionMetrics[name])[0] : 'ULP';
+  const tolerance = getPrecisonTolerance(name, metricType, resources);
+  const expected = resources.expected;
+  let outputData;
+  let expectedData;
+  if (Array.isArray(expected)) {
+    // check outputs by split or gru
     for (let operandName in namedOutputOperands) {
-      let outputData = outputs[operandName];
-      let expectedData = resources.expected.data[operandName];
-      if (typeof expectedData === 'number') {
-        outputData = [outputData];
-        expectedData = [expectedData];
-      }
-      assert_array_approx_equals_ulp(outputData, expectedData, tolerance, operandType, description);
+      outputData = outputs[operandName];
+      expectedData = getExpectedData(expected, operandName);
+      doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
     }
-  } else if (metricType === 'ATOL') {
-    for (let operandName in namedOutputOperands) {
-      assert_array_approx_equals(
-        outputs[operandName], resources.expected.data[operandName], tolerance, description);
+  } else {
+    const outputName = expected.name ? expected.name : 'output';
+    outputData = outputs[outputName];
+    expectedData = expected.data;
+    doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
+  }
+};
+
+/**
+ * Create input operands for a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {String[]} nameArray
+ * @returns {MLOperand[]} Input operands
+ */
+const createInputOperands = (builder, resources, nameArray) => {
```

Comment:
Should this method be used by concat test? It is not used by anyone.

Reply:
Since the inputs of `concat` is a sequence, it's special to other ops.
I form its inputs, like

```
      "inputs": [
        {
          "name": "input1",
          "shape": [12],
          "data": [
            -0.39444134019222243,
            ...
          ]
        },
        {
          "name": "input2",
          "shape": [12],
          "data": [
            0.4918989118791477,
            ....
          ]
        }
      ],
```
while for inputs of `clamp`, I form it like
```
      "inputs": {
        "x": { // use "x" for input operand name
            "shape": [12],
            "data": [
              -0.39444134019222243,
              ...
          -  ]
          }
        }
```
so this `createInputOperands` is for other ops, please see the usage on [demo tests](https://github.com/BruceDai/wpt/pull/442/files) .

Reply:
If this function is not used in this PR, I'd suggest to leave it to following PRs where it is used.